### PR TITLE
[8.x] Add key/index/iteration placeholders to validation attributes

### PR DESCRIPTION
**Pull request by amir9480**
This pull request provides 3 placeholders for validation attributes/messages when validating arrays
and replaces them for better validation messages.

`:key`: which is array key.
`:index`: which is the same key but replaces only numbers (not strings).
`:iteration`: which is the same index plus one (which more readable and useful for the final user).

An example code before this change:
```php
// Sample request ['items' => [4, 8, 'test', 3]]
public function test(Request $request)
{
	  $attributes = [];
      if(is_array($request->input('items'))) {
	      foreach ($request->input('items') as $key => $value) {
	          $attributes["items.$key"] = 'Item ('.(intval($key) + 1).')';
	      }
	  }
	  Validator::make($request->all(), 
          ['items.*' => 'required|numeric|min:1'],
          [],
          $attributes
      )->validate();
      // The Item (3) must be a number.
}
```

Same code after this change:
```php
// Sample request ['items' => [4, 8, 'test', 3]]
public function test(Request $request)
{
	  Validator::make($request->all(), 
          ['items.*' => 'required|numeric|min:1'],
          [],
          ['items.*' => 'Item (:iteration)']
      )->validate();
	  // The Item (3) must be a number.
}
```

---

There are placeholders for nested arrays too and can be handle by adding an index number at end of the placeholder.

```php
Validator::make(
    ['name' => [['', 'Foo'], ['', 'Bar']]],
    ['name.*.*' => 'required'],
    [],
    ['name.*.*' => 'Name (:iteration0)(:iteration1)']
)->validate();
// Error message 1: The Name (1)(1) field is required.
// Error message 2: The Name (2)(1) field is required.
```

### CODE REVIEW

**Comment by taylorotwell** [comment]
Why iterate over the same thing three times?

**Comment by amir9480** [comment]
@taylorotwell 
There is no specific reason for that.
I just tried to write clean and readable code.

What name should I use to cover (key/index/iteration) at the same time?

**Comment by amir9480** [comment]
@taylorotwell 

Is `replaceKeysPlaceholders` fine for all of them?

**Comment by taylorotwell** [comment]
Maybe something like replaceArrayPlaceholders.

File: src/Illuminate/Validation/Concerns/FormatsMessages.php [hunk 1]
```diff
@@ -295,6 +296,53 @@ protected function replaceAttributePlaceholder($message, $value)
...
+    /**
+     * Replace the array attribute :key{n}/:index{n}/:iteration{n} placeholders in the given message.
+     *
+     * @param  string  $message
+     * @param  string  $attribute
+     * @return string
+     */
+    protected function replaceArrayPlaceholders($message, $attribute)
+    {
+        $explicitKeys = $this->getExplicitKeys($attribute);
```
Comment (amir9480) [inline src/Illuminate/Validation/Concerns/FormatsMessages.php:308]: I'm not sure if this line is a breaking change and I should check for method existence or make a pull request for the 9.x branch instead.

**Comment by taylorotwell** [comment]
I dunno, I think I'm going to skip this one. I would just make a wildcard language line that says something like "All entires require a name" instead of having something like "The name (2)(1) field is required."

**Comment by amir9480** [comment]
@taylorotwell 

I know I can use something like this.
But when the number of entries is a lot, it's harder for the final user to detect what the problem is.
Also `required` rule is easier to detect and fix but something like `unique` or `exists` is harder and a message like `All entries must be valid.` is not useful and using an attribute name including the row of field can help more.

```
The name (2)(1) field is required.
```
The above message is just a sample created by me for tests and can be replaced with more readable attributes/messages instead of `Name (:iteration0)(:iteration1)` based on the UI that we using for our application.

I provide more examples, hopefully, I don't need to use the old way I using now to approach this functionality anymore.

Another example:

```php
// Sample request ['items' => [4, 8, 'test', 3]]
public function test(Request $request)
{
	  Validator::make($request->all(), 
          ['items.*' => 'required|numeric|min:1'],
          [],
          ['items.*' => 'Item in row :iteration']
      )->validate();
	  // The Item in row 3 must be a number.
}
```

---

```php
Validator::make(
    ['name' => [['', 'Foo'], ['', 'Bar']]],
    ['name.*.*' => 'required'],
    ['name.*.*.required' => 'The field in row :iteration0 and column :iteration1 is required.']
)->validate();
// Error message 1: The field in row 1 and column 1 is required.
// Error message 2: The field in row 2 and column 1 is required.
```

**Pull request closed without merging**